Citation explorer

adomasven opened the pull request:
![image](https://github.com/zotero/zotero/assets/5899315/b096c2ae-3d5c-45bb-aeba-a927c6be5bf0)

Adds a Citation Explorer dialog to the Word integration as per https://github.com/zotero/zotero/issues/21#issuecomment-1473852674 with the following functionality:

- Find all citations that cite a specific item
- List items in a selected citation
- Highlight a selected citation in the document
- Open up the citation dialog for the selected citation
- Show item in Zotero library (where you can edit item metadat)
- Highlight unlinked items and allow relinking them
- Show a diff for items with modified text
- A tab to see items only cited in the bibliography

Notes:
- A user might use this dialog to remove citations from their document, which depending on the word processor might break citation highlighting. We detect those cases and notify the user that further highlighting of citations is disabled until the dialog is reopened.
- A button has been added to support this new functionality for all word processors. Things are getting kind of busy there with the recent addition of Add Note and now this. We could consider rolling Unlink Citations into Document Preferences, or do something else, but we are quite limited in our options.
- Show in Document works better if we are able to auto-focus the document processor window. In my testing this works worst/not-at-all on Windows, and I actually added a function to activate Word (which, apparently, wasn't previously implemented). Users will have to manually switch to Word on Windows.
- Winword template will need to be signed at the end of this by Dan.

adomasven commented:
Merging this would allow [forum users](https://forums.zotero.org/discussion/111879/quellen-aus-bestehendem-dokument-ins-zotero-uebertragen) to rebuild their library from existing documents.

![image](https://github.com/zotero/zotero/assets/5899315/8ebca859-d7de-4beb-8725-3723f29feb8f)

dstillman commented:
https://forums.zotero.org/discussion/comment/471834/#Comment_471834

abaevbog commented:
I am noticing that with `regularOnly: true`, arrowRight/left can still display/hide child rows with notes and attachments when multiple rows regular items are selected:

https://github.com/user-attachments/assets/27d24449-0bb1-4249-aa9d-6243786c5295

adomasven commented:
> I am noticing that with `regularOnly: true`, arrowRight/left can still display/hide child rows with notes and attachments when multiple rows regular items are selected

It may be so, but I'm not able to reproduce it in the citation dialog. The `reguralOnly` prop wasn't added by this commit, merely moved from being an attribute of the ItemTree object to a prop. Maybe some adjustments in the citation dialog tree are causing it?

abaevbog commented:
> The reguralOnly prop wasn't added by this commit, merely moved from being an attribute of the ItemTree object to a prop.

Oh, I see what you mean now. `regularOnly` has been around for a while since https://github.com/zotero/zotero/commit/cbbff600a60c9e7a7407d6f2e4053309bf28b872, it was just not used, which is why we didn't run into it  before. If I add `this.regularOnly = true` to the constructor of the `itemTree.js`, I get the same behavior in the main itemTree of ZoteroPane. So I'll make a separate issue for it, as it isn't caused by anything citations-related.

adomasven commented:
> If I add `this.regularOnly = true` to the constructor of the `itemTree.js`, I get the same behavior in the main itemTree of ZoteroPane. So I'll make a separate issue for it, as it isn't caused by anything citations-related.

Right, it's actually even older than that, and transferred from the old XUL item tree.

dstillman commented:
Merged hidden in https://github.com/zotero/zotero/commit/a3654652b21071955ae01ba8fcdc38765e1740a4 as part of #4872, but leaving this open for review.

adam3smith commented:
Is this accessible in the current beta (with a hidden pref or a JS call or so)? I'm very curious to try this.

adomasven committed:
Add a hidden pref to enable citation explorer on google docs

adomasven committed:
Restore split menu button fix (d96afa44, regression b2ad17d6)

adomasven committed:
Fix Citation Explorer breaking when library item data not loaded

adomasven committed:
fx128: Remove Service.jsm load in Citation Explorer

adomasven committed:
Fix Citation Explorer changes not saving when switching to citation dialog

adomasven committed:
Fix split menu button not flex-growing

adomasven commented:
This was broken, so updated to make it work again and add a hidden pref `integration.googleDocsCitationExplorerEnabled` to enable it in Google Docs. Enabling in other doc editors is a matter of updating UI. The dotm files and libreoffice changes are also available.

This is what it looks like now:
![image](https://github.com/user-attachments/assets/368fc848-6093-4512-b055-93deb46382f7)

I assume @dstillman won't let this go through before @yexingsha gives their input and standardizes the look. Please have a go at it when possible.

adomasven committed:
Restore tick and cross icons for citation link status

yexingsha commented:
<img width="880" alt="citation-explorer" src="https://github.com/user-attachments/assets/9026d5da-b9aa-431b-90e8-8acf6318c9b3" />

[Figma link](https://www.figma.com/design/y8aZzEDwsxsgVh7Zui63to/Zotero?node-id=14125-128652&t=6REyd79Hf2j0eAWA-1) (I should be able to give you access to this file)

Aside from the obvious changes:
- Making the central divider draggable is probably overkill, but we can split the two areas 40/60 to utilize the space better.
- I think we need placeholder text in search bars; if these are wrong, feel free to correct them.
- Changed the icon of the "Is Linked" column to `link`, since the original icon means attachment.
- I think it makes more sense to put the "Is Linked" column in the same place, rather than mirrored.
- In item tree, "Year" is probably more useful than "Date".
- Changed deleted text to `accent-red`.

Other issues I found while testing:
- Colored tags are missing from titles, and the emoji tags are displayed too tight.
- When I select a citation X, then select an item Y, I don't think the previous selection of citation X and its item highlights should persist. It's busy and confusing, and also makes it difficult to tell whether item Y is cited in citation X, since X is now always colored gray no matter which item you select. The same is true in the opposite direction.

adomasven committed:
Citation Explorer: Update per new design

Switch the layout to use grid instead of flexbox

adomasven committed:
Citation explorer: macOS styling

- Refactor alternating row colors from item tree into virtualized table
- Adjust the way twisty and spacer are aligned for regularOnly item
  trees

adomasven committed:
Citation explorer: Windows styling

yexingsha commented:
Things I'm noticing from the screenshots:

- The icon in the "Add to ..." button should be colored `accent-blue`.
- The gap between "Relink Item" and "Add to ..." should be 8px.
- On macOS, the outer paddings of the two item trees look about 2px smaller than their inner paddings.
- On windows, the divider between sections look weirdly dark.

And there are still outstanding issues from my last comment:

- In item tree, "Year" is probably more useful than "Date".
- Changed deleted text to accent-red.
- Colored tags are missing from titles, and the emoji tags are displayed too tight.
- When I select a citation X, then select an item Y, I don't think the previous selection of citation X and its item highlights should persist. It's busy and confusing, and also makes it difficult to tell whether item Y is cited in citation X, since X is now always colored gray no matter which item you select. The same is true in the opposite direction.

adomasven committed:
Fix switching to different citation explorer view not resetting selection

adomasven committed:
Citation dialog: Fix tab order

adomasven committed:
Fix splitMenuButton appearance on Linux

adomasven committed:
Citation Dialog: log failures to select citation in document

adomasven committed:
Citation Explorer: Fix item tree menu view

adomasven committed:
Citation Explorer: Change Date column out for Year

adomasven committed:
Citation Explorer: Various styling adjustments

adomasven commented:
MacOS:
<img width="914" height="640" alt="image" src="https://github.com/user-attachments/assets/a053db8f-f50d-4249-a5c6-fd433cf1cf25" />

adomasven commented:
Windows:

<img width="802" height="532" alt="Untitled" src="https://github.com/user-attachments/assets/1b9a4300-f31b-41e8-8bb0-d2a7caa6f117" />

Thanks all for the comments!

> 1. When a row in the left ("Citation") table is selected and selection changes in the right itemTree, I am seeing a noticeable flicker of the row that the selected citation has. In the recording below, the item "A Candid Assessment ..." does the flickering:
> * I think the tab order in the dialog at the moment is not quite right. It spans the entire document in a way that doesn't really correspond to the layout. I think that as you press Tab, the focus should move in the following order, similar to ZoteroPane: `#citation-search` -> `#citation-list` -> `#button-show-in-document` -> `#button-edit-citation` -> `cited/uncited tabbox` -> `#item-search` -> `itemTree` -> `#button-show-in-zotero` -> `#button-relink-item` -> `#button-addTo-library` -> `OK`
> * The View menu with options to sort or rearrange rows of the itemTree (for a11y) sometimes does not appear for me when the citation explorer opens. I think I was running into a similar issue with citation dialog and [fixed it by adding ](https://github.com/zotero/zotero/blob/e27b58176d45b2cba4f44d08248ec70d1ffc55e6/chrome/content/zotero/integration/citationDialog.xhtml#L49-L50) the `<xul:item-tree-menu-bar/>` node into the `.xhtml` file instead of waiting for itemTree to create it.

> * The icon in the "Add to ..." button should be colored `accent-blue`.
> * The gap between "Relink Item" and "Add to ..." should be 8px.
> * On windows, the divider between sections look weirdly dark.
> * In item tree, "Year" is probably more useful than "Date".
> * Changed deleted text to accent-red.
> * Colored tags are missing from titles, and the emoji tags are displayed too tight.
> * When I select a citation X, then select an item Y, I don't think the previous selection of citation X and its item highlights should persist. It's busy and confusing, and also makes it difficult to tell whether item Y is cited in citation X, since X is now always colored gray no matter which item you select. The same is true in the opposite direction.

Fixed. Some of these were supposed to be addressed, but the code didn't quite work or regressed with other changes.

@abaevbog:

> 2. If i click on the "Show in Document", I always just get a warning saying "Citations in your document have been modified since Citation Explorer has been opened...", and when I close it, the "Show in Document" and "Edit Citation" buttons are both disabled, and nothing else happens. Is that expected?

This is due to a bug in Google Docs editor that will be fixed. It will work if you first place a cursor in a citation when opening the Citation Explorer, or after closing and reopening after it didn't work at first. I've explained why this exists in the initial PR comment.

> 3. Would it make sense to disable "Edit Citation" when multiple citations are selected? It currently opens the first selected citation but, since many are selected, it implies that somehow all selected citations will be opened for edit, I think. We could also consider to move that button as a column of the Citations table.
> 4. On the other hand, would we want to enable "Show in Zotero" button if multiple items are selected in the itemTree? I understand that the items are very likely to not be anywhere near each other. But it seems like people are often interested in fetching all cited items to group or export them. So this would allow one to get all cited items selected in their Library, after which one can move them into a collection for further exporting/processing/etc.

This is more of a design consideration. Overall, probably the most reasonable thing to do is to disable multi-select in these trees. There are two reasons for multi-select: being able to highlight multiple items in the other tree by selecting multiple in one (in case you're troubleshooting where which citation is included, or what have you), the other is multiple Show in Zotero, which you correctly observe, currently doesn't work.

Overall, I'm not sure if it's good UX to have buttons that appear or disappear based on context, so in that sense maybe buttons in the tree would be better. The other option is having buttons that are disabled and enabled based on context. I'm not a very big fan of buttons in the tree for this, since there are all quite discrete and contextual actions that are better explained with text than an icon.


@yexingsha:

> * On macOS, the outer paddings of the two item trees look about 2px smaller than their inner paddings.

Not sure what you mean here.

abaevbog commented:
> This is more of a design consideration. Overall, probably the most reasonable thing to do is to disable multi-select in these trees. There are two reasons for multi-select: being able to highlight multiple items in the other tree by selecting multiple in one (in case you're troubleshooting where which citation is included, or what have you), the other is multiple Show in Zotero, which you correctly observe, currently doesn't work.

With these two use cases, multi-select feels quite important here. One should be able to find out which items are in a set of citations, right? And being able to find a selection of items in Zotero library is a necessary step to be able to, say, place all of your cited items into a collection before exporting them somewhere (which I believe is a frequently asked about use case). 

> Overall, I'm not sure if it's good UX to have buttons that appear or disappear based on context, so in that sense maybe buttons in the tree would be better. The other option is having buttons that are disabled and enabled based on context.

I agree that hiding/showing buttons depending on context is a bit jumpy. Disabling/enabling buttons instead could feel smoother.

Placing buttons into the columns is harder here than in other instances. For example, in citation dialog, we have `+` buttons, and it's quite self-explanatory what it would do. I think some buttons here are better candidates than others to be placed into the tables - for instance, "Edit citation" would probably make sense with a standard "Edit" icon. But others, like "Show in Document" or "Relink Item" in itemTree - not so much.

The benefit of buttons in the table is that you have only one click needed to complete the action vs two (selection + button click). But not sure if it's worth it here.

abaevbog commented:
> Right. Since uncited items are always added from Edit Bibliography, let's use the same label system from there:

I certainly like this look without tabs. And it's nice to draw a parallel with the Edit Bibliography dialog too. The only question I have is how would one find all uncited items? If one has many citations, it could be challenging to scroll through a long list. Is that a valid concern? If so, an alternative options could be to place the "Uncited" label into a separate column.- then one would be able to sort by it.

yexingsha commented:
> And being able to find a selection of items in Zotero library is a necessary step to be able to, say, place all of your cited items into a collection before exporting them somewhere (which I believe is a frequently asked about use case).

Oh, this is a good point. Then I guess we do need to allow multi-select and fix "Show in Zotero." I think disabling the "Edit Citation" button when multiple items are selected should be fine; maybe we can add a tooltip to the button in this state that says "You can only edit one citation at a time." Or something like that.

> The only question I have is how would one find all uncited items? If one has many citations, it could be challenging to scroll through a long list. Is that a valid concern?

I guess people might try to find uncited items if e.g. they added some entries to the bibliography, but then found out they need to either cite or delete them because of requirements of a journal or their professor. Since you can't actually delete a bibliography entry in the citation explorer, and I don't think it makes sense to enable it here (since we are looking at items, not really bibliography entries), this seems more like a concern for the Edit Bibliography dialog.

adomasven committed:
Citation Explorer: Fix highlighting inconsistencies

adomasven commented:
> > This is more of a design consideration. Overall, probably the most reasonable thing to do is to disable multi-select in these trees. There are two reasons for multi-select: being able to highlight multiple items in the other tree by selecting multiple in one (in case you're troubleshooting where which citation is included, or what have you), the other is multiple Show in Zotero, which you correctly observe, currently doesn't work.
> 
> With these two use cases, multi-select feels quite important here. One should be able to find out which items are in a set of citations, right? And being able to find a selection of items in Zotero library is a necessary step to be able to, say, place all of your cited items into a collection before exporting them somewhere (which I believe is a frequently asked about use case).

Yeah, wanting to select all items and add to a collection in Zotero is a valid use case and users will want to do that. There's some caveats here:

1. The more obvious one: if you select multiple items with some unlinked to Zotero, then Show in Zotero won't have a target to highlight for those items. That's probably fine and not incredibly unexpected as a UI behaviour.
2. Less obvious: if the document contains items cited from multiple libraries, and you press Show in Zotero, the behaviour is undefined.

One unorthodox solution for 2 would be to add a temporary virtual collection in Zotero when Citation Explorer dialog is open, and highlight items there. Then, make that collection in Zotero contain more information, in the item pane, showing whether the item is linked, and from which library, and perhaps with some columns too, so that you can sort on them. The virtual collection disappears on citation explorer close.

adomasven committed:
Citation Explorer: Uncited item display change

adomasven committed:
Citation Explorer: minor styling changes

adomasven committed:
Citation Explorer: Disable buttons when no selection

abaevbog commented:
> I've made a version which disables buttons unless there is a selection, and I think it's bad. I'd rather the buttons only appear when there's an action for a corresponding item.

Yeah, it looks a bit busy on your screenshot... I think on mac it looks better, as long as the window is not shrunk to the min width. Personally, I do prefer this version with disabled buttons because I find it helpful to always see what options are available. For instance, as soon as the dialog opens for the first time, I immediately see that I can "Edit a citation", as opposed to discovering it later and having to somewhat remember that it's one of the options. But maybe that's just my personal perference. @yexingsha, do you have any thoughts on the above?

<img width="865" height="531" alt="Screenshot 2025-09-29 at 9 38 29 AM" src="https://github.com/user-attachments/assets/3dd1b45b-2a90-4ffc-b6f9-78fd1001b0d0" />

> Upon more thinking, I think we should make the Add to Collection button always enabled, for single and multiple selection, that way the ability to save all items into a collection is right there in the dialog.

Yes, I fully agree with that. What do you think about also having the "Export" button? "Add to Collection" naturally won't work for items across different libraries but one might need to get the full list of items sometimes? With this, we may be running into the issue of having just too many buttons squashed next to one another. What do you think about moving the buttons relevant for multiple selected items into a `menupopup` as `menuitems`? Then we could replace most of the buttons below itemTree ("Add to Collection", "Export", "Show in Zotero") with a single `toolbarbutton` labelled "Options" or "Actions", or something like that?

> The only remaining question is what to do with items that are not from the same library. The simplest solution is to add another column "Library" and only show it when there are items cited from multiple libraries 

Yeah, disabling the buttons when items from different collections are selected sounds reasonable to me. The column would work fine functionally. I think it may not be immediately clear that some options get disabled because multiple selected items are in different libraries - one would have to kinda discover it. We can remedy that by adding a tooltip to relevant buttons explaining why they are disabled. Another drawback is that space is somewhat limited here, so it feels a bit wasteful to have a column that will often have repeated information. Instead of the column, we could do the grouping of items by their libraries (as we tried briefly in the [annotations pr](https://github.com/zotero/zotero/pull/5356#issuecomment-3051089977)). That comes with various challenges of reworking the itemTree though and is probably just an overkill here.

yexingsha commented:
The UI looks mostly fine now, except that:

- On macOS, the buttons now have too little bottom margin.
- When the "Add to..." button is disabled, the icon should use `accent-blue50`.

I still prefer disabling the buttons instead of hiding them, for the same reasons @abaevbog outlined. And disabled buttons can provide tooltips explaining why they are disabled.

As to how to handle actions on items from different libraries, I agree that adding a "Library" column is probably not that helpful. Even with this column, I assume most people will only figure out they can't show/add items from different libraries *after* they have multiple items selected and see the tooltip on a disabled button, and at that point it will still be a hassle to modify their selection. So I think grouping items by libraries, as we tried in the annotations PR, is actually pretty useful here. Since this dialog does not have a collectionTree, the issues we encountered there no longer apply. And it's less wasteful than a whole column.

If we do add "Export", then I agree it's a good idea to put these actions under the same "Actions..." button. But I wonder what's the use case for "Export" in this dialog, and how many people would use it?

abaevbog commented:
> If we do add "Export", then I agree it's a good idea to put these actions under the same "Actions..." button. But I wonder what's the use case for "Export" in this dialog, and how many people would use it?

Well, without it, if one has cited items across 5 different groups, in order to export a full list of citations, one would have to select all items for each of those 5 libraries, export them one library at a time in the main window, and then put those parts together, right? And that sounds like a lot of unnecessary work for the user. Here, we can help them export the entire list in one go.

> So I think grouping items by libraries, as we tried in the annotations PR, is actually pretty useful here.

@adomasven, the itemTree sections setup I put together in the annotations PR was pretty raw, since it was just an experiment (that we didn't end up going for). But if it would be helpful, I can see if any of those parts could be useful here. Let me know!

abaevbog commented:
I meant to mention that all logic for copying items between libraries currently is in `collectionTree`, and even then it's not quite complete (e.g. if linked items exist but their metadata is different, there's no way to easily fix it).

In https://github.com/zotero/zotero/pull/5577/, among other things, I pulled the copying of items into dedicated methods, like `Zotero.Items.copyToLibrary`, which could come in handy for copying items between groups.

adomasven commented:
Finally, some updates! After heavily refactoring item tree so that we can support other things than zotero items as item roots, we now have sectioned items:

<img width="1900" height="1178" alt="image" src="https://github.com/user-attachments/assets/5b773ee3-dd6f-4cb6-a719-33d30ff1c2d2" />

Moreover, a wizard for all your item adjustment needs!

<img width="1252" height="861" alt="image" src="https://github.com/user-attachments/assets/632548da-7203-4b30-a02f-b92eb1b9de3b" />

<img width="1252" height="861" alt="image" src="https://github.com/user-attachments/assets/f6996f3c-71fb-441c-b6f5-730f743594c8" />

And an item relink picker with suggestions:

<img width="1820" height="1078" alt="image" src="https://github.com/user-attachments/assets/24dd7f36-49dd-4c04-8e4b-6c6b9a6e90c2" />

@abaevbog @yexingsha @dstillman I believe this is now in a mergable state. You can test it in Google Docs, where Citation Explorer menu entry will appear upon detection of this build.

abaevbog commented:
Very exciting! This works great!

The only issue I have ran into so far is the "Show in Document" button does not work for me in a fresh document. I keep getting the warning that citations have been changed in my document since when the explorer was opened.

One other thought is if there should be any additional handling at the relinking step of an unlinked item? Suppose I link a cited item to an item in my library but the item in Zotero has a somewhat different metadata. It can significantly change the citations from what they were before - should we warn about it in some way?

I'll keep playing around to see if I have anything else!
